File: src/account_disp.ml
let make_withdraw
    (balance : int ref) : int -> result =
  fun amount ->
    if !balance >= amount
    then (balance := !balance - amount;
          Ok !balance)
    else Err "Insufficient funds"

let () =
  let w1 = make_withdraw (ref 100)
  and w2 = make_withdraw (ref 100)
  in
  print_result (w1 10); (* Ok(90) *)
  print_result (w1 20); (* Ok(70) *)
  print_result (w1 80); (* Err("Insufficient funds") *)
  
  print_result (w2 40)  (* Ok(60) *)

File: src/account_disp.rs
fn make_withdraw(mut balance: i32) ->
    Box<FnMut(i32) -> Result<i32, &'static str>> {
    Box::new(move |amount: i32| {
        if balance >= amount {
            balance -= amount;
            Ok(balance)
        } else { Err("Insufficient funds") }
    })
}
let mut w1 = make_withdraw(100);
let mut w2 = make_withdraw(100);

println!("{:?}", w1(10)); // Ok(90)
println!("{:?}", w1(20)); // Ok(70)
println!("{:?}", w1(80)); // Err("Insufficient funds")

println!("{:?}", w2(40)); // Ok(60)
